sympy/functions/combinatorial/numbers.py

```diff
@@ -1715,22 +1715,25 @@ def _stirling1(n, k):
         return S.One
     if 0 in (n, k):
         return S.Zero
-    n1 = n - 1
 
     # some special values
     if n == k:
         return S.One
-    elif k == 1:
-        return factorial(n1)
-    elif k == n1:
+    elif k == n - 1:
         return binomial(n, 2)
     elif k == n - 2:
         return (3*n - 1)*binomial(n, 3)/4
     elif k == n - 3:
         return binomial(n, 2)*binomial(n, 4)
 
-    # general recurrence
-    return n1*_stirling1(n1, k) + _stirling1(n1, k - 1)
+    row = [1]+[0 for _ in range(k)]
+    for i in range(1, n+1):
+        new = [0]
+        for j in range(1, k+1):
+            stirling = (i-1) * row[j] + row[j-1]
+            new.append(stirling)
```

Comment:
Is it possible to reuse `row` (instead of appending to `new`) if `j` would run from `k` down to `1`?

Reply:
I don't think so, as far as I can understand, you are suggesting a Top-Down approach, where we will try to reuse row, but don't know whether the same relation will hold.

Reply:
I suggest `row[j] = (i-1) * row[j] + row[j-1]`.

Reply:
> I suggest `row[j] = (i-1) * row[j] + row[j-1]`.
Just give me sometime, will try and let you know.

Reply:
> I suggest `row[j] = (i-1) * row[j] + row[j-1]`.

Some tests are failing after doing this change.

Reply:
>  if j would run from k down to 1

Together with this change?

Reply:
> > if j would run from k down to 1
> 
> Together with this change?

for j in range(K+1, 1) ??

Reply:
`for j in range(k, 0, -1)` should do that.

Reply:
Yes they are failing.

Reply:
Please review the pull request @oscarbenjamin @sylee957 @jksuom 
If any changes are still required, please suggest, otherwise I think the request is ready to be merged.

Reply:
The tests succeed on my system with this diff:
```
--- a/sympy/functions/combinatorial/numbers.py
+++ b/sympy/functions/combinatorial/numbers.py
@@ -1726,13 +1726,10 @@ def _stirling1(n, k):
     elif k == n - 3:
         return binomial(n, 2)*binomial(n, 4)
 
-    row = [1]+[0 for _ in range(k)]
-    for i in range(1, n+1):
-        new = [0]
-        for j in range(1, k+1):
-            stirling = (i-1) * row[j] + row[j-1]
-            new.append(stirling)
-        row = new
+    row = [0, 1] + [0]*(k-1)  # for n = 1 
+    for i in range(2, n+1):
+        for j in range(k, 0, -1):
+            row[j] = (i-1) * row[j] + row[j-1]
     return Integer(row[k])
```
The initial value `row[0] = 1` is only needed for computing the case `i = 1`. It is necessary to replace `row[0]` by 0 afterwards as is done here: `new = [0]`. This could be set explicitly after the `j`-loop. However, I would prefer starting from the case `n = 1` where `row[0] = 0` is already set. (The case `n = 0` has been handled above.)

Reply:
That's true, yeah these are small optimizations which should have been made. Extremely sorry for the inconvenience.

Reply:
Is there any issue with _stirling2 as well?

Reply:
Another optimization is possible. For small values of `i`, `row` will mostly consist of zeros. It is not necessary to reassign them for `j > i`. So it seems that the `j` loop could be `for j in range(min(k, i), 0, -1)`.

Reply:
Any further improvements after this one? @jksuom @sylee957 @oscarbenjamin 
If there are, please suggest. Otherwise, I will commit all the changes once and for all, so that the request gets ready to be merged.